File: Platforms/Windows/Reign.Video.D3D9/BufferLayout.cpp
#include "pch.h"
#include "BufferLayout.h"

namespace Reign
{namespace Video
{namespace D3D9
{
	#pragma region Constructors
	BufferLayout::BufferLayout(DisposableI^ parent, ShaderI^ shader, BufferLayoutDescI^ bufferLayoutDesc)
	: Disposable(video)
	{
		null();
		try
		{
			video = parent->FindParentOrSelfWithException<Video^>();

			IDirect3DVertexDeclaration9* layoutTEMP;
			BufferLayoutDesc^ bufferLayoutDescTEMP = (BufferLayoutDesc^)bufferLayoutDesc;
			if (FAILED(video->Device->CreateVertexDeclaration(bufferLayoutDescTEMP->Desc, &layoutTEMP)))
			{
				layout = 0;
				Debug::ThrowError(L"BufferLayout", L"Failed to create VertexDeclaration");
			}
			layout = layoutTEMP;
		}
		catch (Exception^ ex)
		{
			delete this;
			throw ex;
		}
	}

	BufferLayout::~BufferLayout()
	{
		disposeChilderen();
		if (layout) layout->Release();
		null();
	}

	void BufferLayout::null()
	{
		layout = 0;
		video = nullptr;
	}
	#pragma endregion
	
	#pragma region Methods
	void BufferLayout::Enable()
	{
		video->Device->SetVertexDeclaration(layout);
	}
	#pragma endregion
}
}
}

File: Platforms/Vita/Reign.Video.Vita/Buffers/BufferLayout.cs
using Reign.Core;
using System;

namespace Reign.Video.Vita
{
	public class BufferLayout : Disposable, BufferLayoutI
	{
		#region Properties
		private Video video;
		internal BufferLayoutDesc desc;
		#endregion

		#region Constructors
		public static BufferLayout New(DisposableI parent, ShaderI shader, BufferLayoutDescI desc)
		{
			return new BufferLayout(parent, shader, desc);
		}

		public BufferLayout(DisposableI parent, ShaderI shader, BufferLayoutDescI desc)
		: base(parent)
		{
			video = parent.FindParentOrSelfWithException<Video>();
			this.desc = (BufferLayoutDesc)desc;
		}
		#endregion

		#region Methods
		public void Enable()
		{
			// Not used in Vita
		}
		#endregion
	}
}